### PR TITLE
[E2E] Device tracking

### PR DESCRIPTION
**Pull request by Zil0**
This is the most complicated part (yet?). It holds up in production, but I probably haven't experienced every race conditions that can occur.

Problems can also arise when persisting the list and shutting down unexpectedly. For reference, here is the commit that shows how persistence is done in device_list.py https://github.com/matrix-org/matrix-python-sdk/pull/243/commits/fbe5164eac5742c03a00e7731b2b93978cd99e61#diff-e1d31eea9af27e3baaf26050b334d1ab.

Signed-off-by: Valentin Deniaud \<valentin.deniaud@inpt.fr\>

### CODE REVIEW

File: matrix_client/client.py [hunk 1]
```diff
@@ -604,9 +604,15 @@ def _sync(self, timeout_ms=30000):
             # TODO: the rest of this for loop should be in room object method
             room.prev_batch = sync_room["timeline"]["prev_batch"]
 
+            new_users = set()
```
Comment (Zil0) [inline matrix_client/client.py]: I was reflecting on this thinking it was ugly, and I noticed that it was also completely broken (doesn't look at the join events in `timeline`). But, this has been there for a month with everything working perfectly in all of my tests. In fact I think this is simply unneeded as the `changed` key of `device_lists` of the sync response already tells us everything we need to know ie it is populated the first time we share an encrypted room with a user. If whoever reviews this agrees, I'll drop this commit.

**Comment by Zil0** [comment]
@non-Jedi the problem to solve here is to download the keys of a user whenever they join an encrypted room, and track their device list for changes. Those happen very often, and we are notified via a new field in `/sync/`. The whole download process is also quite heavy with a lot of checks. For these reasons, this all has to be done in a non-blocking way. Also, we have to ensure that only one download request is in flight at a time in order to prevent various races from happening.
Hence this PR adds a background thread, which will always be there if encryption is enabled when instantiating `MatrixClient`.  
You may like to comment on the general idea, but I don't think it could have been done any other way (mostly I felt that you should know about this thread and the rationale behind it). What gets added to [client.py](https://github.com/matrix-org/matrix-python-sdk/pull/238/files#diff-6770c150117812114aea5a8aab1b7da9) and [room.py](https://github.com/matrix-org/matrix-python-sdk/pull/238/files#diff-1e3f0d25ba4b2cf8ef4f9434e7538be0) is pretty minimal and straightforward, since all of the logic is hidden in `device_list.py`.

**Comment by non-Jedi** [comment]
I haven't dug into this yet, but for my info, is this using the stuff described at https://github.com/matrix-org/matrix-doc/issues/1212

**Comment by Zil0** [comment]
Yes, but strangely enough this part has made it into the spec https://matrix.org/docs/spec/client_server/unstable.html#tracking-the-device-list-for-a-user, except for the small addition of the `left` parameter in sync response.

File: matrix_client/crypto/device_list.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    def sync_token(self):
+        return self._sync_token
+
+    @sync_token.setter
+    def sync_token(self, token):
+        if not self._sync_token or token > self._sync_token:
+            self._sync_token = token
+
+
+class UpdateDeviceList(Thread):
```
Comment (non-Jedi) [inline matrix_client/crypto/device_list.py]: I would prefer this be called `UpdateDeviceListThread` or similar.

File: matrix_client/crypto/device_list.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        Args:
+            room (Room): The room to use.
+            blocking (bool): Optional. Whether to wait for the keys to have been
+                downloaded before returning.
+        """
+        logger.info('Fetching all missing keys in room %s.', room.room_id)
+        user_ids = {u.user_id for u in room.get_joined_members()} - self.tracked_user_ids
+        if not user_ids:
+            logger.info('Already had all the keys in room %s.', room.room_id)
+            if blocking:
```
Comment (non-Jedi) [inline matrix_client/crypto/device_list.py]: What is this waiting on in this case? As far as I can tell, no work has been given to the thread when it reaches this.

File: matrix_client/crypto/device_list.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+                # Wait on an eventual download to finish
+                self.update_thread.event.wait()
+            return
+        with self.thread_condition:
```
Comment (non-Jedi) [inline matrix_client/crypto/device_list.py]: So this wakes up the `UpdateDeviceList.run` method, right, when it calls `self.thread_condition.notify()`? All of this threading code is really hard to follow primarily because the python stdlib interface to threading appears to be awful and woefully underdocumented in docstrings.

To remedy this and make this stuff maintainable, I think more comments are needed anywhere the code touches a `Condition` or `Event`. For one thing, both the `Event` and `Condition` attached to the thread are used for control flow, but I'm still not sure what meaning they're each supposed to convey.

File: matrix_client/crypto/device_list.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+    @sync_token.setter
+    def sync_token(self, token):
+        if not self._sync_token or token > self._sync_token:
+            self._sync_token = token
+
+
+class UpdateDeviceList(Thread):
+
+    def __init__(self, cond, user_ids, download_method, tracked_user_ids):
```
Comment (non-Jedi) [inline matrix_client/crypto/device_list.py]: Up above, I talk about confusion wrt threading code, I think some comments here describing how each of `self.cond`, `self.event`, and `self.should_terminate` is used would help clear up a lot of my confusion.

File: matrix_client/client.py [hunk 6]
```diff
@@ -579,6 +579,15 @@ def _mkroom(self, room_id):
     # TODO better handling of the blocking I/O caused by update_one_time_key_counts
     def _sync(self, timeout_ms=30000):
         response = self.api.sync(self.sync_token, timeout_ms, filter=self.sync_filter)
+
+        if self._encryption and 'device_lists' in response:
+            if response['device_lists'].get('changed'):
```
[the remark this answers is not in the document]
Comment (Zil0) [reply matrix_client/client.py]: They won't do the exact same thing, as `if response['device_lists'].get('changed')` will evaluate to `False` if the response is like `{'device_list': {'changed': []}}`, which IIRC it tends to be.

File: matrix_client/crypto/device_list.py (hunk 2 above)
Comment (Zil0) [reply matrix_client/crypto/device_list.py]: This is only for internal use, hence I don't think the name is that important. Since I agree it appears very confusing in the generated doc (even though Sphinx tells you it inherits from `Thread`), I've already changed it to `_UpdateDeviceList`.

File: matrix_client/crypto/device_list.py (hunk 3 above)
Comment (Zil0) [reply matrix_client/crypto/device_list.py]: Consider:
```
device_list.get_room_device_keys(room)
device_list.get_room_device_keys(room)
```
Without this check, the second call may return, even though the keys are still in the process of being retrieved (since we flag users as tracked before having successfully downloaded their keys).

File: matrix_client/crypto/device_list.py (hunk 4 above)
Comment (Zil0) [reply matrix_client/crypto/device_list.py]: Yes, this is used to wake up the thread. I think the interface is explained pretty well in the online doc https://docs.python.org/3/library/threading.html#condition-objects.

**Comment by Zil0** [comment]
I will squash the various fixups already present in this PR and try to add more comments.

**Comment by Zil0** [comment]
Done, let me know if it helps.

File: matrix_client/client.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (non-Jedi) [reply matrix_client/client.py]: ah gotcha
